src/esp/gfx/PbrDrawable.cpp

```diff
@@ -0,0 +1,162 @@
+                    materialData_->roughnessTexture,
+                    materialData_->metallicTexture,
+                    materialData_->noneRoughnessMetallicTexture,
+                    materialData_->occlusionRoughnessMetallicTexture,
+                    materialData_->normalTexture)
+      .setBaseColor(materialData_->baseColor)
+      .setRoughness(materialData_->roughness)
+      .setMetallic(materialData_->metallic);
+  if (materialData_->textureMatrix != Mn::Matrix3{}) {
+    shader_->setTextureMatrix(materialData_->textureMatrix);
+  }
+
+  shader_->draw(mesh_);
+}
+
+Mn::ResourceKey PbrDrawable::getShaderKey(Mn::UnsignedInt lightCount,
+                                          PbrShader::Flags flags) const {
+  return Corrade::Utility::formatString(
+      SHADER_KEY_TEMPLATE, lightCount,
+      static_cast<PbrShader::Flags::UnderlyingType>(flags));
+}
+
+PbrDrawable& PbrDrawable::updateShader() {
+  PbrShader::Flags flags = PbrShader::Flag::ObjectId;
+
+  if (materialData_->textureMatrix != Mn::Matrix3{}) {
+    flags |= PbrShader::Flag::TextureTransformation;
+  }
+
+  if (materialData_->baseColorTexture) {
+    flags |= PbrShader::Flag::BaseColorTexture;
+  }
+  if (materialData_->occlusionRoughnessMetallicTexture) {
+    flags |= PbrShader::Flag::OcclusionRoughnessMetallicTexture;
+  }
+  if (materialData_->noneRoughnessMetallicTexture) {
+    flags |= PbrShader::Flag::NoneRoughnessMetallicTexture;
+  }
+  if (materialData_->roughnessTexture)
+    flags |= PbrShader::Flag::RoughnessTexture;
+  if (materialData_->metallicTexture)
+    flags |= PbrShader::Flag::MetallicTexture;
+  if (materialData_->normalTexture)
+    flags |= PbrShader::Flag::NormalTexture;
+  if (materialData_->perVertexObjectId) {
+    // TODO: may be supported in the future
+  }
+
+  // correct the flags if necessary
+  flags = PbrShader::generateCorrectFlags(flags);
+
+  unsigned int lightCount = lightSetup_->size();
+  if (!shader_ || shader_->lightCount() != lightCount ||
+      shader_->flags() != flags) {
+    // if the number of lights or flags have changed, we need to fetch a
+    // compatible shader
+    shader_ = shaderManager_.get<Mn::GL::AbstractShaderProgram, PbrShader>(
+        getShaderKey(lightCount, flags));
+
+    // if no shader with desired number of lights and flags exists, create one
+    if (!shader_) {
+      shaderManager_.set<Mn::GL::AbstractShaderProgram>(
+          shader_.key(), new PbrShader{flags, lightCount},
+          Mn::ResourceDataState::Final, Mn::ResourcePolicy::ReferenceCounted);
+    }
+
+    CORRADE_INTERNAL_ASSERT(shader_ && shader_->lightCount() == lightCount &&
+                            shader_->flags() == flags);
+  }
+
+  return *this;
+}
+
+// update every light's color, intensity, range etc.
+PbrDrawable& PbrDrawable::updateShaderLightParameters() {
+  constexpr float dummyRange = Mn::Constants::inf();
+  std::vector<float> ranges;
+  std::vector<Mn::Color3> colors;
```

Comment:
```suggestion
  std::vector<Mn::Color3> colors;
  colors.reserve(lightSetup_->size());
```

Like you do in the other function below.

Reply:
OK